design/design.md

```diff
@@ -0,0 +1,203 @@
+# CDM Spark Events Design
+
+## Nomenclature
+
+* CTS - the [CDM Task Service](https://github.com/kbase/cdm-task-service)
+
+## Purpose
+
+The CDM Spark events repo is responsible for automatically taking action based on upstream events.
+The initial target is processing output from the CTS, but additional event sources
+could be added in the future, e.g. responding to data uploaded to a dataplane bucket.
+
+The action taken depends on the nature of the event. For instance, if a CTS job completes and
+places job output in the data plane, the event system may run code to import data into CDM
+Deltalake tables.
+
+## Notes on Spark Structured Streaming
+
+Since the initial target of the event processor is importing CTS job data into the Spark
+Deltalake tables, an obvious choice of framework for the processor would be
+Spark Structured Streaming (SSS). However, after an investigation, it seems SSS is ill-suited
+to our needs:
+
+* Since an HTTP request must be made to the CTS to fetch the job data, including the output files,
+  the SSS data sink must allow for arbitrary code. This
+  [limits the sinks](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#output-sinks)
+  to the `forEach`   and `forEachBatch` sinks.
+* `forEach` code
+  [runs on the Spark cluster worker](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/ForeachWriter.html#process-T-).
+  This means that the only the resources of that worker, not the entire cluster, can be used to
+  process the job output data. As such `forEach` is not an acceptable choice.
+* `forEachBatch` processes data on the driver, and so the driver code can submit full cluster
+  jobs for the CTS jobs in the batch. However,
+  [only one batch runs at a time](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#triggers)
+  which means the processing time for the batch is equal to the processing time of largest
+  job output in the batch if the batch is parallelized. `forEachBatch` doesn't allow for
+  true parallelism - instead, from a performance perspective, it effectively picks the largest job
```

Comment:
job => jobs

Reply:
I think job is correct here, it's talking about a singular job

Reply:
The sentence is:

> it effectively picks the largest job out of the N jobs in the batch, and sequentially runs those large jobs.

"picks the largest job" / "runs those large jobs"

Reply:
Tried to clarify the text a bit though

Reply:
> "picks the largest job" / "runs those large jobs"

Right, it picks the largest job from each batch, and then runs those large jobs, 1 per batch

Reply:
Thanks, that's clearer.